frontend/src/metabase/nav/hooks.ts

```diff
@@ -0,0 +1,19 @@
+import { useDatabaseListQuery } from "metabase/common/hooks";
+import { useSelector } from "metabase/lib/redux";
+import { PLUGIN_SELECTORS } from "metabase/plugins";
+import { getIsPaidPlan } from "metabase/selectors/settings";
+import { getUserIsAdmin } from "metabase/selectors/user";
+
+export function useShouldShowDatabasePromptBanner(): boolean | undefined {
```

Comment:
oh, I see, I didn't notice `onlyHaveSampleDatabase` can be undefined, such name presumes we have a boolean value inside IMO

Reply:
what if we explicitly return `undefined` in database is not available yet? @kamilmielnik @WiNloSt

Reply:
what about renaming it to `maybeOnlyHaveSampleDatabase`? I think I've seen this pattern somewhere.

Reply:
I presume it can break falsy statements 😄

Reply:
> what if we explicitly return `undefined` if database is not available yet? @kamilmielnik @WiNloSt

even better! I like that.

Reply:
@uladzimirdev  The best I can come up with is
```diff
diff --git a/frontend/src/metabase/nav/hooks.ts b/frontend/src/metabase/nav/hooks.ts
index db3c7436d1..f13d7ebb31 100644
--- a/frontend/src/metabase/nav/hooks.ts
+++ b/frontend/src/metabase/nav/hooks.ts
@@ -12,8 +12,13 @@ export function useShouldShowDatabasePromptBanner(): boolean | undefined {
   const { data: databases } = useDatabaseListQuery({
     enabled: isAdmin && isPaidPlan && !isWhiteLabeling,
   });
+
+  if (isAdmin && isPaidPlan && !isWhiteLabeling && databases == null) {
+    return undefined;
+  }
+
   const onlyHaveSampleDatabase =
-    databases && databases.length === 1 && databases[0].is_sample;
+    databases?.length === 1 && databases[0].is_sample;
 
   return isAdmin && isPaidPlan && !isWhiteLabeling && onlyHaveSampleDatabase;
 }
```

Or the full function
```ts
export function useShouldShowDatabasePromptBanner(): boolean | undefined {
  const isAdmin = useSelector(getUserIsAdmin);
  const isPaidPlan = useSelector(getIsPaidPlan);
  const isWhiteLabeling = useSelector(PLUGIN_SELECTORS.getIsWhiteLabeling);

  const { data: databases } = useDatabaseListQuery({
    enabled: isAdmin && isPaidPlan && !isWhiteLabeling,
  });

  if (isAdmin && isPaidPlan && !isWhiteLabeling && databases == null) {
    return undefined;
  }

  const onlyHaveSampleDatabase =
    databases?.length === 1 && databases[0].is_sample;

  return isAdmin && isPaidPlan && !isWhiteLabeling && onlyHaveSampleDatabase;
}
```

What do you think?

Reply:
am I right, that we care about databases only if all other conditions are truthy? @WiNloSt

Reply:
@uladzimirdev yes. The trick is that the original statement `isAdmin && isPaidPlan && !isWhiteLabeling && onlyHaveSampleDatabase` which `onlyHaveSampleDatabase` could be undefined will not care about `onlyHaveSampleDatabase` if the conditions before that return `false` aka shorting the circuit.

You could also think that we don't actually need to know if `onlyHaveSampleDatabase` is true or not if other conditions return `false` since `onlyHaveSampleDatabase` rely on an API call. That's why the logic is a little hairy hear. To make sure we only fetch the API only when needed.

Reply:
I think the usage of `null` is confusing as we're not dealing with it.

My take:
```tsx
export function useShouldShowDatabasePromptBanner(): boolean | undefined {
  const isAdmin = useSelector(getUserIsAdmin);
  const isPaidPlan = useSelector(getIsPaidPlan);
  const isWhiteLabeling = useSelector(PLUGIN_SELECTORS.getIsWhiteLabeling);
  const isEligibleForPromptBanner = isAdmin && isPaidPlan && !isWhiteLabeling;

  const { data: databases } = useDatabaseListQuery({
    enabled: isEligibleForPromptBanner,
  });

  if (!isEligibleForPromptBanner) {
    return false;
  }

  if (typeof databases === 'undefined') {
    return undefined;
  }

  const onlyHasSampleDatabase = databases.length === 1 && databases[0].is_sample;
  return onlyHasSampleDatabase;
}
```

Reply:
On paper, I think @kamilmielnik 's suggestion is really easy to follow what we will return at each condition, but it makes the whole condition when we will render the banner less obvious than just
`isAdmin && isPaidPlan && !isWhiteLabeling && onlyHaveSampleDatabase`

This is a trade-off what do you  guys think?

Reply:
@kamilmielnik's version is better, I didn't notice the same condition in the `useDatabaseListQuery` 👏

Reply:
anecdotally, @uladzimirdev 's take is the first one I've got (but never commented here)

Reply:
@WiNloSt nope, it's crystal clear when to return what in the latest versions. Otherwise I needed to ask you 😄

Reply:
@uladzimirdev I understand that. But it's less obvious what the actual condition that the product care which is `isAdmin && isPaidPlan && !isWhiteLabeling && onlyHaveSampleDatabase`

vs having to read through a few more lines of code. But I'll take that. I think that's more obvious at least for engineers how each parameter matters.

Reply:
as engineer, I don't know the business requirement and based on the condition `isAdmin && somethingThatCanBeUndefined` it's not clear what was the product decision about it.

With more explicit version it's clear: if we didn't meet requirements (isAdmin ...) - we return false and do not show a banner.

If we met requirements, let's check if it's the only database.

if the database is not loaded yet, we return undefined. 

Probably for the last case it's worth having an API like `{loading: false, data: ..., error: undefined}`

Reply:
I've seen the idea that exposing the loading state 2 times now. But let's keep things simple for now. To callers, they don't need to care whether the result will be `false` or `undefined` since they won't render the banner anyway. But don't forget the reason `undefined` exist in the first place is to make unit testing `useShouldShowDatabasePromptBanner` simpler to avoid having to check if the API is called for certain conditions before asserting.

Reply:
Fixed using everyone's input aka @kamilmielnik 's suggestion with some modification.